<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Yan Li, Michiko Rye, Qing Gong" />

<meta name="date" content="2021-07-27" />

<title>scRICA: single-cell RNA-Seq Integrative Comparative Analysis</title>

<script src="data:application/javascript;base64,Ly8gUGFuZG9jIDIuOSBhZGRzIGF0dHJpYnV0ZXMgb24gYm90aCBoZWFkZXIgYW5kIGRpdi4gV2UgcmVtb3ZlIHRoZSBmb3JtZXIgKHRvCi8vIGJlIGNvbXBhdGlibGUgd2l0aCB0aGUgYmVoYXZpb3Igb2YgUGFuZG9jIDwgMi44KS4KZG9jdW1lbnQuYWRkRXZlbnRMaXN0ZW5lcignRE9NQ29udGVudExvYWRlZCcsIGZ1bmN0aW9uKGUpIHsKICB2YXIgaHMgPSBkb2N1bWVudC5xdWVyeVNlbGVjdG9yQWxsKCJkaXYuc2VjdGlvbltjbGFzcyo9J2xldmVsJ10gPiA6Zmlyc3QtY2hpbGQiKTsKICB2YXIgaSwgaCwgYTsKICBmb3IgKGkgPSAwOyBpIDwgaHMubGVuZ3RoOyBpKyspIHsKICAgIGggPSBoc1tpXTsKICAgIGlmICghL15oWzEtNl0kL2kudGVzdChoLnRhZ05hbWUpKSBjb250aW51ZTsgIC8vIGl0IHNob3VsZCBiZSBhIGhlYWRlciBoMS1oNgogICAgYSA9IGguYXR0cmlidXRlczsKICAgIHdoaWxlIChhLmxlbmd0aCA+IDApIGgucmVtb3ZlQXR0cmlidXRlKGFbMF0ubmFtZSk7CiAgfQp9KTsK"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="data:text/css,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">scRICA: <strong>s</strong>ingle-<strong>c</strong>ell <strong>R</strong>NA-Seq <strong>I</strong>ntegrative <strong>C</strong>omparative <strong>A</strong>nalysis</h1>
<h4 class="author">Yan Li, Michiko Rye, Qing Gong</h4>
<h4 class="date">2021-07-27</h4>



<div id="scrica-single-cell-rna-seq-integrative-comparative-analysis" class="section level1">
<h1>scRICA: <strong>s</strong>ingle-<strong>c</strong>ell <strong>R</strong>NA-Seq <strong>I</strong>ntegrative <strong>C</strong>omparative <strong>A</strong>nalysis</h1>
<div id="what-is-scrica" class="section level3">
<h3>1. What is scRICA</h3>
<p>It is a R workflow package which can be used to perform scRNA-Seq downstream integrative, comparative analyses and visualization. This package can process a batch of scRNA-Seq count matrix from different experimental conditions fro integration and comparative analysis efficiently and reproducible. The package functions can be categorized into: 1). analysis workflow functions and 2). visualization functions:</p>
<div id="workflow-functions" class="section level4">
<h4>1.1 Workflow functions:</h4>
<ul>
<li><strong><code>processQC()</code></strong>, process quality control (QC) assessment for all samples defined in a metadata table, including doublets identification with wrapped function <strong><code>findDoublets()</code></strong>, gene features stigmatization, and mitochondrial content filtering.
<ul>
<li><strong><code>findDoublets()</code></strong>, this function is based on <a href="https://github.com/EDePasquale/DoubletDecon">DoubletDecon</a> to perform either or both medoids and centroids deconvolution algorithms (<a href="https://www.cell.com/cell-reports/fulltext/S2211-1247(19)31286-0" class="uri">https://www.cell.com/cell-reports/fulltext/S2211-1247(19)31286-0</a>) defined in the metadata table for doublet cells detection.</li>
</ul></li>
<li><strong><code>getIntegrationClusterMarkers()</code></strong>, 1). perform integrative analysis based on cells passing the QC with function <strong><code>processQC()</code></strong>; 2). conduct cell clustering analyses to identify different clustering cell types; and 3). identify conserved gene markers from each identified cell clusters.</li>
<li><strong><code>getExpCondClusterMarkers()</code></strong>, identify top expressed gene markers conserved in samples from different experimental conditions.</li>
<li><strong><code>getClusterExpCondDe()</code></strong>, identify differential expressed gene markers from 2 specified different experimental condition samples.</li>
<li><strong><code>getExpCondClusterPseudotime()</code></strong>, perform pseudotime trajectory functional analysis on the specified experimental condition clustering cells with 3 different methods, including principal components analysis (PCA) based on <a href="https://bioconductor.org/packages/release/bioc/html/scater.html">Scater</a>, Diffusion Maps based on <a href="https://bioconductor.org/packages/release/bioc/html/destiny.html">density</a>, and <a href="https://www.bioconductor.org/packages/release/bioc/html/slingshot.html">slingshot</a>.</li>
</ul>
</div>
<div id="visualization-functions" class="section level4">
<h4>1.2 Visualization functions:</h4>
<ul>
<li><strong><code>getClusterSummaryReplot()</code></strong>: 1). summarize the number of cell in each identified or annotated clustering cell types; and 2). generate corresponding tSNE and UMAP plots on the specified experimental conditions.</li>
<li><strong><code>getGoiDotplot()</code></strong>: generate dot-plots of provided marker genes on the specified experimental conditions.</li>
<li><strong><code>getGoiFeatureplot()</code></strong>: generate feature-plots of provided marker genes on the specified experimental conditions.</li>
<li><strong><code>plotPseudotime()</code></strong>: generate pseudotime functional trajectory plots from 3 different analysis methods.</li>
<li><strong><code>plotPseudotimeLineages()</code></strong>: make pseudotime trajectory lineage plots based on slingshot.</li>
<li><strong><code>plotPseudotimeHeatmap()</code></strong>: make heat-map of identified genes differentially expressed on the pseudotime trajectory lineages from slingshot.</li>
</ul>
</div>
</div>
<div id="pacakge-installation" class="section level3">
<h3>2. pacakge installation</h3>
<ul>
<li><p>github installation</p>
<pre><code>devtools::install_github(repo = &#39;yan-cri/scRICA&#39;, build_vignettes = T, force = T)
library(scRICA)</code></pre></li>
<li><p>local download installation Download package to your local computer via <code>git clone https://github.com/yan-cri/scRICA.git</code></p>
<pre><code>devtools::install(&#39;scRICA&#39;, build_vignettes = T)
library(scRICA)</code></pre></li>
<li><p>The package usage information can be seen from package vignettes via</p>
<pre><code>browseVignettes(package = &#39;scRICA&#39;)</code></pre></li>
</ul>
</div>
<div id="input-metadata-table-file" class="section level3">
<h3>3. Input metadata table file</h3>
<p>This workflow package has its own inherited structure, and requires an initial metadata table to initiate the entire scRNA-Seq workflow analyses. 4 columns are required in the metadata table, they are <code>sample</code>, <code>path</code>, <code>doubletsRmMethod</code> and <code>expCond1</code> for samples names, full path to sample’s count matrix tables directory (cellranger analysis results), doublets detective methods with four options, and experimental condition levels respectively; up to 2 experimental conditions specified in column <code>expCond1</code> and <code>expCond2</code> can be explored with this package. If no doublets removal is needed for samples, please specify ‘none’ for that sample in the column <code>doubletsRmMethod</code>.</p>
<table>
<colgroup>
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
<col width="20%" />
</colgroup>
<thead>
<tr class="header">
<th>sample</th>
<th>path</th>
<th>expCond1</th>
<th>expCond2</th>
<th>doubletsRmMethod</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>sample1_condA_cond1</td>
<td>/FullPath/to/CountMatrix/</td>
<td>condA</td>
<td>cond1</td>
<td>OL/centroids/medoids/none</td>
</tr>
<tr class="even">
<td>sample2_condA_cond2</td>
<td>/FullPath/to/CountMatrix/</td>
<td>condA</td>
<td>cond2</td>
<td>OL/centroids/medoids/none</td>
</tr>
<tr class="odd">
<td>sample3_condA_cond3</td>
<td>/FullPath/to/CountMatrix/</td>
<td>condA</td>
<td>cond3</td>
<td>OL/centroids/medoids/none</td>
</tr>
<tr class="even">
<td>sample4_condB_cond1</td>
<td>/FullPath/to/CountMatrix/</td>
<td>condB</td>
<td>cond1</td>
<td>OL/centroids/medoids/none</td>
</tr>
<tr class="odd">
<td>sample5_condB_cond2</td>
<td>/FullPath/to/CountMatrix/</td>
<td>condB</td>
<td>cond2</td>
<td>OL/centroids/medoids/none</td>
</tr>
</tbody>
</table>
</div>
<div id="analysis-workflow-implementations-demonstration-data" class="section level3">
<h3>4. Analysis workflow implementations demonstration data</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">read.delim2</span>(<span class="at">file =</span> <span class="fu">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;metadata_mac.txt&#39;</span>, <span class="at">package =</span> <span class="st">&#39;scRICA&#39;</span>, <span class="at">mustWork =</span> T), <span class="at">header =</span> T) </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>metadata</span></code></pre></div>
<pre><code>##   sample expCond1 expCond2
## 1  A3396     3396        A
## 2  F3396     3396        F
## 3  I3396     3396        I
## 4  A3041     3041        A
## 5  I3041     3041        I
##                                                                                   path
## 1 /Library/Frameworks/R.framework/Versions/4.0/Resources/library/scRICA/extdata/3396A/
## 2 /Library/Frameworks/R.framework/Versions/4.0/Resources/library/scRICA/extdata/3396F/
## 3 /Library/Frameworks/R.framework/Versions/4.0/Resources/library/scRICA/extdata/3396I/
## 4  /Library/Frameworks/R.framework/Versions/4.0/Resources/library/scRICA/extdata/3041A
## 5  /Library/Frameworks/R.framework/Versions/4.0/Resources/library/scRICA/extdata/3041I
##   doubletsRmMethod
## 1               OL
## 2               OL
## 3               OL
## 4             none
## 5             None</code></pre>
</div>
<div id="analysis-workflow-implementations" class="section level3">
<h3>5. Analysis workflow implementations</h3>
<p>This package separate the scRNA-Seq integrative and comparative analysis into 3 categories, including integrative analysis with QC assessment for doublets and mitochondrial content, comparative analysis from different experimental condition groups, and visualizations.</p>
<div id="integrative-analysis-with-qc-assessment" class="section level4">
<h4>5.1 integrative analysis with QC assessment</h4>
<ul>
<li>QC results without mitochondrial content filtering</li>
</ul>
<pre><code>library(scRICA)
?processQC  
qcResult1 &lt;- processQC(metadata = metadata, 
                     resDirName = &#39;scRICA_test_result1&#39;, 
                     genomeSpecies = &#39;human&#39;)</code></pre>
<ul>
<li>QC results with mitochondrial content filtering: turn on <code>mtFiltering</code> option and setup the mitochondrial content percentage cut-off values with option <code>mtPerCutoff</code></li>
</ul>
<pre><code>qcResult2 &lt;- processQC(metadata = metadata, 
                       resDirName = &#39;scRICA_test_result2&#39;, 
                       genomeSpecies = &#39;human&#39;, 
                       mtFiltering = T, mtPerCutoff = 20)</code></pre>
</div>
<div id="experimental-condition-comparative-analysis" class="section level4">
<h4>5.2 experimental condition comparative analysis</h4>
<ul>
<li></li>
</ul>
</div>
<div id="visualizations" class="section level4">
<h4>5.3 visualizations</h4>
<ul>
<li></li>
</ul>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
