---
title: "scRICA: **s**ingle-**c**ell **R**NA-Seq **I**ntegrative **C**omparative **A**nalysis"
output: rmarkdown::html_vignette
date: "`r Sys.Date()`"
author: "Yan Li, Michiko Rye, Qing Gong"
vignette: >
  %\VignetteIndexEntry{scRICA_v1}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# scRICA: **s**ingle-**c**ell **R**NA-Seq **I**ntegrative **C**omparative **A**nalysis 

### 1. What is scRICA
It is a R workflow package which can be used to perform scRNA-Seq downstream integrative, comparative analyses and visualization. This package can process a batch of scRNA-Seq count matrix from different experimental conditions fro integration and comparative analysis efficiently and reproducible. The package functions can be categorized into: 1). analysis workflow functions and 2). visualization functions:

#### 1.1 Workflow functions:

  * **`processQC()`**, process quality control (QC) assessment for all samples defined in a metadata table, including doublets identification with wrapped function **`findDoublets()`**, gene features stigmatization, and mitochondrial content filtering.
    + **`findDoublets()`**, this function is based on [DoubletDecon](https://github.com/EDePasquale/DoubletDecon) to perform either or both medoids and centroids deconvolution algorithms (https://www.cell.com/cell-reports/fulltext/S2211-1247(19)31286-0) defined in the metadata table for doublet cells detection.
  * **`getIntegrationClusterMarkers()`**, 1). perform integrative analysis based on cells passing the QC with function **`processQC()`**; 2). conduct cell clustering analyses to identify different clustering cell types; and 3). identify conserved gene markers from each identified cell clusters.
  * **`getExpCondClusterMarkers()`**, identify top expressed gene markers conserved in samples from different experimental conditions.
  * **`getClusterExpCondDe()`**, identify differential expressed gene markers from 2 specified different experimental condition samples.
  * **`getExpCondClusterPseudotime()`**, perform pseudo time trajectory functional analysis on the specified experimental condition clustering cells with 3 different methods, including  principal components analysis (PCA) based on [Scater](https://bioconductor.org/packages/release/bioc/html/scater.html), Diffusion Maps based on [density](https://bioconductor.org/packages/release/bioc/html/destiny.html), and [slingshot](https://www.bioconductor.org/packages/release/bioc/html/slingshot.html).

#### 1.2 Visualization functions:

  * **`getClusterSummaryReplot()`**: 1). summarize the number of cell in each identified or annotated clustering cell types; and 2). generate corresponding tSNE and UMAP plots on the specified experimental conditions.
  * **`getGoiDotplot()`**: generate dot-plots of provided marker genes on the specified experimental conditions.
  * **`getGoiFeatureplot()`**: generate feature-plots of provided marker genes on the specified experimental conditions.
  * **`plotPseudotime()`**: generate pseudo time functional trajectory plots from 3 different analysis methods.
  * **`plotPseudotimeLineages()`**: make pseudo time trajectory lineage plots based on slingshot.
  * **`plotPseudotimeHeatmap()`**: make heat-map of identified genes differential expressed on the pseudo time trajectory lineages from slingshot.

### 2. pacakge installation

  * github installation
    ```
devtools::install_github(repo = 'yan-cri/scRICA', build_vignettes = T, force = T)
library(scRICA)
```

  * local download installation
    Download package to your local computer via ``git clone https://github.com/yan-cri/scRICA.git``
    ```
devtools::install('scRICA', build_vignettes = T)
library(scRICA)
```
  * The package usage information can be seen from package vignettes via 
    ```
browseVignettes(package = 'scRICA')
```
  
### 3. Input metadata table file

This workflow package has its own inherited structure, and requires an initial metadata table to initiate the entire scRNA-Seq workflow analyses. 4 columns are required in the metadata table, they are `sample`, `path`, `doubletsRmMethod` and `expCond1` for samples names, full path to sample's count matrix tables directory (cellranger analysis results), doublets detective methods with four options, and experimental condition levels respectively; up to 2 experimental conditions specified in column `expCond1` and `expCond2` can be explored with this package. If no doublets removal is needed for samples, please specify 'none' for that sample in the column `doubletsRmMethod`.

sample  | path  | expCond1 | expCond2 | doubletsRmMethod
------------- | -------------  | -------------  | ------------- | ------------- 
sample1_condA_cond1  | /FullPath/to/CountMatrix/ | condA | cond1 | OL/centroids/medoids/none 
sample2_condA_cond2  | /FullPath/to/CountMatrix/ | condA | cond2 | OL/centroids/medoids/none 
sample3_condA_cond3  | /FullPath/to/CountMatrix/ | condA | cond3 | OL/centroids/medoids/none
sample4_condB_cond1  | /FullPath/to/CountMatrix/ | condB | cond1 | OL/centroids/medoids/none 
sample5_condB_cond2  | /FullPath/to/CountMatrix/ | condB | cond2 | OL/centroids/medoids/none 

### 4. Analysis workflow implementations demonstration data 
```{r echo=TRUE}
system.file('extdata', package = 'scRICA', mustWork = T)
metadata <- read.delim2(file = system.file('extdata', 'metadata_mac.txt', package = 'scRICA', mustWork = T), header = T) 
metadata
```

### 5. Analysis workflow implementations

This package separate the scRNA-Seq integrative and comparative analysis into 3 categories, including integrative analysis with QC assessment for doublets and mitochondrial content, comparative analysis from different experimental condition groups, and visualizations.


#### 5.1 integrative analysis with QC assessment
  * QC assessment
    + QC assessment without mitochondrial content filtering
    
      ```
library(scRICA)
qcResult1 <- processQC(metadata = metadata, resDirName = 'scRICA_test_result1', genomeSpecies = 'human')
```
    + QC assessment without mitochondrial content filtering: turn on `mtFiltering` option and setup the mitochondrial content percentage cut-off values with option `mtPerCutoff`
    
      ```
library(scRICA)
qcResult <- processQC(metadata = metadata, resDirName = 'scRICA_test_result', genomeSpecies = 'human', mtFiltering = T, mtPerCutoff = 20)
```
      
      ```
## -- scRICA_test_result
##    |__doublet_results
##       |__ ...
##       |__doubletDecon_preProcessed_results
##          |__ ...
##    |__MT_percentage_summary.txt
##    |__No_filtered_cells_summary.txt
##    |__org_doubletsRemoval_cellNoSummary.txt
##    |__QC_plots
##       |__featureScatter_A3041.pdf
##       |__ ...
##       |__featureViolin_A3041.pdf
##       |__ ...
##       |__topVariableFeature_A3041.pdf
##       |__ ...
```
    The analyses results will be saved in the defined `resDirName = 'scRICA_test_result'` option under the current working directory shown as below data structure, where includes 2 directories and 3 text files. 
      ```{r echo=F}
source('/Users/yanli/Desktop/scRIC_development/scRICA_fnPseudotime_exploration/twee.R')
##twee(path = '/Users/yanli/Desktop/scRIC_development/scRICA_test_result_vig/') 
```
      + Directory 'doublet_results' include the doublets detection results;
      + Directory 'QC_plots' contains all quality assessement results with respect to each sample in the metadata table;
      + Text file 'MT_percentage_summary.txt' presents a summary of mitochondrial content persentage for all samples;
      + Text file 'org_doubletsRemoval_cellNoSummary.txt' shows a summary of number of cells before and after doublets removal;
      + Text file 'No_filtered_cells_summary.txt' presents number of cells before and after mitochondrial content filtering, if otpion ` mtFiltering = T`.
    
  * Integration analyses
    ```
results <- getClusterMarkers(qcProcessedResults = qcResult)
```

    ```
## -- scRICA_test_result
##    |__allCluster_pos_markers_no.txt
##    |__allCluster_pos_markers_top10.txt
##    |__allCluster_pos_markers.txt
##    |__cluster_heatmap_top10PosMarkers.pdf
##    |__pca_plot.pdf
##    |__RDS_Dir
##       |__scRICA_test_result.rds
##    |__tsne_plot_samplSep.pdf
##    |__tsne_plot.pdf
##    |__umap_plot_samplSep.pdf
##    |__umap_plot.pdf
```

    The integration results will be saved in the same dirctory defined in the `processQC()` execution. In addition to the previous `processQC()`, we can see that the integration analysis results are shown as below data structure.

    + Integration analysis results are saved in RDS file inside directory 'RDS_Dir';
    + Two types of clustering results are saved in file name starting with 'tsne*' and 'umap*' respectively;
    + 'allCluster_pos_markers.txt' includes the identified gene markers with resepct to each cell cluster;
    + 'allCluster_pos_markers_no.txt' summarize the number of identified gene markers in each cell cluster;
    + 'allCluster_pos_markers_top10.txt' presents the top N (defined in the option `topN` of `getClusterMarkers()`) identified gene markers with resepct to each cell cluster;
    + 'cluster_heatmap_top10PosMarkers.pdf' shows a heatmap of the top N (defined in the option `topN` of `getClusterMarkers()`) identified gene markers from each cell cluster.

  * Integrated results summarization and visualization
    
    The integrated analyses result can be summarized on the different experimental conditions with function `getClusterSummaryReplot()` shown as below, where user can define which experimental conditions to be summarized. If user would like to add cluster cells annotation, this can be doen by turn on the option `newAnnotation = T` and provide correpsonding Rscript name with option `newAnnotationRscriptName` wher the corresponding cluster cells annotations are defined.

      ```
getClusterSummaryReplot(resDir = results$resDir, newAnnotation = F, expCondCheck = 'sample', expCondSepName = 'sample_org')
```

      ```
## -- scRICA_test_result
##    |__results_wOrgClusterAnnotation
##       |__expCond_sample_org
##          |__cellNo_summary_orgClusterAnnotation_expCond_sample_org.pdf
##          |__cellNo_summary_orgClusterAnnotation_expCond_sample_org.txt
##          |__new_tSNE_plot_expCond_sample_org
##             |__tsne_plot_noLabel_integrate_orgAnnotation.pdf
##             |__tsne_plot_wLabel_integrate_orgAnnotation.pdf
##             |__tsne_plot_wLabel_orgAnnotation_expCond_sample_org.pdf
##          |__new_UMAP_plot_expCond_sample_org
##             |__UMAP_plot_noLabel_integrate_orgAnnotation.pdf
##             |__UMAP_plot_wLabel_integrate_orgAnnotation.pdf
##             |__UMAP_plot_wLabel_orgAnnotation_expCond_sample_org.pdf
```
    
    The option `expCondCheck` of `sample`, `expCond1`, or `expCond2` summarized the identified clusters cell numbers with respect to each sample defined in the metadata table column `sample`, experimental condition levels defined in the metadata table column `expCond1` or `expCond2` respectively. The summarized results are saved inside the directory 'results_wOrgClusterAnnotation' or 'results_wOrgClusterAnnotation' under the directory name of option `resDir`.
    
      ```
getClusterSummaryReplot(resDir = results$resDir, newAnnotation = F, expCondCheck = 'expCond1', expCondSepName = 'expCond1_org' )
getClusterSummaryReplot(resDir = results$resDir, newAnnotation = F, expCondCheck = 'expCond2', expCondSepName = 'expCond2_org' )
```

      ```
## -- scRICA_test_result
##    |__results_wOrgClusterAnnotation
##       |__expCond_expCond1_org
##          |__ ...
##       |__expCond_expCond2_org
##          |__ ...
##       |__expCond_sample_org
##          |__ ...
```


#### 5.2 markers genes explorations
  
  * dot plot exploration
  
    ```   
getGoiDotplot(resDir = results$resDir, goiFname = 'scRICA/misc/marker_genes.xlsx', expCondCheck = 'expCond1', expCondSepName = 'expCond1_org')
```

    ```
## -- scRICA_test_result
##    |__results_wOrgClusterAnnotation
##       |__dotplot_selected_markers_expCond_expCond1_org
##          |__goiDotplots_markerGenes_dotplot_expCond_expCond1_org.pdf
```

  * feature plot exploration
    ```   
getGoiFeatureplot(resDir = results$resDir, goiFname = 'scRICA/misc/marker_genes.xlsx', expCondCheck  = 'expCond1')
```

    ```
## -- scRICA_test_result
##    |__results_wOrgClusterAnnotation
##       |__featurePlot_selected_markers_expCond1
##          |__goiFeaturePlot
##             |__umap_expMaxValSummary_expCondexpCond1.txt
##             |__umap_featurePlot_CAPS_expCondexpCond1.pdf
##             |__umap_featurePlot_CD79A_expCondexpCond1.pdf
##             |__ ...
```

    ```   
getGoiFeatureplot(resDir = results$resDir, goiFname = 'scRICA/misc/marker_genes_set2.xlsx', expCondCheck  = 'expCond1', featurePlotFnamePrefix='goiFeaturePlot_set2')
```

    ```
## -- scRICA_test_result
##    |__results_wOrgClusterAnnotation
##       |__featurePlot_selected_markers_expCond1
##          |__goiFeaturePlot
##             |__umap_featurePlot_COL1A1_expCondexpCond1.pdf
##             |__umap_featurePlot_CSPG4_expCondexpCond1.pdf
##             |__ ...
##          |__goiFeaturePlot_set2
##             |__umap_expMaxValSummary_expCondexpCond1.txt
##             |__umap_featurePlot_CAPS_expCondexpCond1.pdf
##             |__ ...
```

#### 5.3 experimental condition comparative analysis
  *

### 6. Supplementary content

The entire scRNA-Seq integrative and comparative analysis can be implemented with below workflow steps based on the input metadata table and other associated options for each step's analysis. User can skip doublet detection step starting from step2 to perform integration analysis directly by setting up metadata table column `'doubletsRmMethod'` as `none`, otherwise, start the analysis from step1 `findDoublets()` with first 2 columns of above metadata table information. 

We provided an ovarian caner study of 1 patient samples from 3 different Fallopian tissue types as this package demonstration data set, which can be accessible at package 'inst/extdata' folder with command `list.files(system.file('extdata', package = 'scRICA', mustWork = T))`. For the demonstration, we also include the corresponding metadata table for this experiment accessible at `system.file('extdata', 'metadata.txt', package = 'scRICA', mustWork = T)`. Please be aware column `path` information in this demonstration metadata table is based on the Mac system, if you are using other operating system, please make your own metadata table.

  * Step1: doublet cells detection: 
```{r}
metadata                <- read.delim2(file = system.file('extdata', 'metadata_mac.txt', package = 'scRICA', mustWork = T), header = T) 
# doubletDeconResDir      <- findDoublets(metadata = metadata, 
#                                         genomeSpecies = 'human', 
#                                         doubletDeconRhop =  0.5, 
#                                         doubletDeconPMF = 'F', 
#                                         resFilename = 'scRICA_test_doublets_checking')
```
This step will conduct cells doublets detection and save the entire analysis results into the defined `resFilename` under current operating directory `getwd()`, it will return a full path of directory to indict where the results were saved at. 

  * Step2: quality control and integrative analysis

This steps analysis will use 2 functions: `countReadin()` and `getClusterMarkers()`.

Based on the previous step's analysis, we need to update our metadata data by including column `doubletsRmMethod` and `doubletsResDir` as shown in above metadata table input file section, where `doubletsRmMethod` allow users to specify which doublet cells detection analysis method to be used for doublet cells filtering, 3 extra options are provided here, 'medoids', 'centroids', or 'OL' (doublet cells detected by both medoids and centroids deconvolution algorithms); `doubletsResDir` specify where the previous analysis results are located. The metadata table can be updated as below and used for this step's integrative analysis via function `countReadin()`.


This function will conduct integrative analysis for all samples provided in the metadata table. If user would like to filter out the mitochondrial content, as shown here, please specify `mtFiltering = 'T'` together with the mitochondrial content filtering percentage with option `mtPerCutoff`. This function will 1) prepare a Seurat object by reading the count matrix from different samples specified in the metadata table into R, and 2) perform the quality control evaluations. 

In addition to integrative analysis for all samples, this function will also conduct the top expressed markers detection with respect to each cell clusters. The entire analysis results will be saved in the specified `resDirName`, the final integrated RDS file will be saved in `'RDS_Dir'` folder inside `resDirName` directory. If no `resDirName` is specified, a folder named as `integration_analysis_results` will be created to save this function's analysis results.

  * Step3: gene marker identfication
2 types of gene markers identification with respect to experimental conditions can be performed with corresponding functions `getClusterExpCondDe()` and `getClusterExpCondDe()`, where as specified above, `getExpCondClusterMarkers()` can identify top expressed gene markers for samples from different experimental conditions with respect to each identified or annotated clustering cell types; and `getClusterExpCondDe()` can identify differential expressed gene markers for samples from 2 different experimental conditions with respect to each identified or annotated clustering cell types.

 Step4: clustering and gene marker visualizations
The integrative cell clustering analysis results can be visualized via function `getClusterSummaryReplot()`, and additional gene markers exploration can be done via function `getGoiDotplot()` and `getGoiFeatureplot()`.

#### 5. Step5: pseudotime functional trajectory analysis and visualization

Pseudotime functional trajectory analysis can be performed at different specified experimental cell clusters via function `getClusterPseudo()`, and the corresponding visualization can be performed via function `plotObjPseudotime()`.

## Feedback
If you have further questions or suggestions regarding this package, please contact Yan Li at yli22@bsd.uchicago.edu from the bioinformatics core at the Center for Research Bioinformatics (CRI), biological science division (BSD), University of Chicago.




