---
title: "scRICA: **s**ingle-**c**ell **R**NA-Seq **I**ntegrative **C**omparative **A**nalysis"
output: rmarkdown::html_vignette
date: "`r Sys.Date()`"
author: "Yan Li, Michiko Rye, Qing Gong"
vignette: >
  %\VignetteIndexEntry{scRICA_v1}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# scRICA: **s**ingle-**c**ell **R**NA-Seq **I**ntegrative **C**omparative **A**nalysis 

### 1. What is scRICA
It is a R workflow package which can be used to perform scRNA-Seq downstream integrative, comparative analyses and visualization. This package can process a batch of scRNA-Seq count matrix from different experimental conditions fro integration and comparative analysis efficiently and reproducible. The package functions can be categorized into: 1). analysis workflow functions and 2). visualization functions:

#### 1.1 Workflow functions:

  * **`processQC()`**, process quality control (QC) assessment for all samples defined in a metadata table, including doublets identification with wrapped function **`findDoublets()`**, gene features stigmatization, and mitochondrial content filtering.
    + **`findDoublets()`**, this function is based on [DoubletDecon](https://github.com/EDePasquale/DoubletDecon) to perform either or both medoids and centroids deconvolution algorithms (https://www.cell.com/cell-reports/fulltext/S2211-1247(19)31286-0) defined in the metadata table for doublet cells detection.
  * **`getIntegrationClusterMarkers()`**, 1). perform integrative analysis based on cells passing the QC with function **`processQC()`**; 2). conduct cell clustering analyses to identify different clustering cell types; and 3). identify conserved gene markers from each identified cell clusters.
  * **`getExpCondClusterMarkers()`**, identify top expressed gene markers conserved in samples from different experimental conditions.
  * **`getClusterExpCondDe()`**, identify differential expressed gene markers from 2 specified different experimental condition samples.
  * **`getExpCondClusterPseudotime()`**, perform pseudotime trajectory functional analysis on the specified experimental condition clustering cells with 3 different methods, including  principal components analysis (PCA) based on [Scater](https://bioconductor.org/packages/release/bioc/html/scater.html), Diffusion Maps based on [density](https://bioconductor.org/packages/release/bioc/html/destiny.html), and [slingshot](https://www.bioconductor.org/packages/release/bioc/html/slingshot.html).

#### 1.2 Visualization functions:

  * **`getClusterSummaryReplot()`**: 1). summarize the number of cell in each identified or annotated clustering cell types; and 2). generate corresponding tSNE and UMAP plots on the specified experimental conditions.
  * **`getGoiDotplot()`**: generate dot-plots of provided marker genes on the specified experimental conditions.
  * **`getGoiFeatureplot()`**: generate feature-plots of provided marker genes on the specified experimental conditions.
  * **`plotPseudotime()`**: generate pseudotime functional trajectory plots from 3 different analysis methods.
  * **`plotPseudotimeLineages()`**: make pseudotime trajectory lineage plots based on slingshot.
  * **`plotPseudotimeHeatmap()`**: make heat-map of identified genes differentially expressed on the pseudotime trajectory lineages from slingshot.

### 2. pacakge installation

  * github installation
    ```
devtools::install_github(repo = 'yan-cri/scRICA', build_vignettes = T, force = T)
library(scRICA)
```

  * local download installation
    Download package to your local computer via ``git clone https://github.com/yan-cri/scRICA.git``
    ```
devtools::install('scRICA', build_vignettes = T)
library(scRICA)
```
  * The package usage information can be seen from package vignettes via 
    ```
browseVignettes(package = 'scRICA')
```
  

### 3. Input metadata table file

This workflow package has its own inherited structure, and requires an initial metadata table to initiate the entire scRNA-Seq workflow analyses. 4 columns are required in the metadata table, they are `sample`, `path`, `doubletsRmMethod` and `expCond1` for samples names, full path to sample's count matrix tables directory (cellranger analysis results), doublets detective methods with four options, and experimental condition levels respectively; up to 2 experimental conditions specified in column `expCond1` and `expCond2` can be explored with this package. If no doublets removal is needed for samples, please specify 'none' for that sample in the column `doubletsRmMethod`.

sample  | path  | expCond1 | expCond2 | doubletsRmMethod
------------- | -------------  | -------------  | ------------- | ------------- 
sample1_condA_cond1  | /FullPath/to/CountMatrix/ | condA | cond1 | OL/centroids/medoids/none 
sample2_condA_cond2  | /FullPath/to/CountMatrix/ | condA | cond2 | OL/centroids/medoids/none 
sample3_condA_cond3  | /FullPath/to/CountMatrix/ | condA | cond3 | OL/centroids/medoids/none
sample4_condB_cond1  | /FullPath/to/CountMatrix/ | condB | cond1 | OL/centroids/medoids/none 
sample5_condB_cond2  | /FullPath/to/CountMatrix/ | condB | cond2 | OL/centroids/medoids/none 

### 4. Analysis workflow implementations demonstration data 
```{r echo=TRUE}
metadata <- read.delim2(file = system.file('extdata', 'metadata_mac.txt', package = 'scRICA', mustWork = T), header = T) 
metadata
```

### 5. Analysis workflow implementations

This package separate the scRNA-Seq integrative and comparative analysis into 3 categories, including integrative analysis with QC assessment for doublets and mitochondrial content, comparative analysis from different experimental condition groups, and visualizations.

#### 5.1 integrative analysis with QC assessment
  * QC results without mitochondrial content filtering
  ```
  library(scRICA)
  ?processQC  
qcResult1 <- processQC(metadata = metadata, 
                       resDirName = 'scRICA_test_result1', 
                       genomeSpecies = 'human')
```
  * QC results with mitochondrial content filtering: turn on `mtFiltering` option and setup the mitochondrial content percentage cut-off values with option `mtPerCutoff`

```
qcResult2 <- processQC(metadata = metadata, 
                       resDirName = 'scRICA_test_result2', 
                       genomeSpecies = 'human', 
                       mtFiltering = T, mtPerCutoff = 20)
```

#### 5.2 experimental condition comparative analysis
  *
  
#### 5.3 visualizations
  *
  




